source/blog/2016/git-git-the-world-s-smallest-git-plugin.html.markdown

```diff
@@ -0,0 +1,155 @@
+---
+date: 2016-09-28 00:48:53 UTC
+title: "git-git: The World's Smallest Git Plugin"
+teaser: >
+  I solved the most minor of my daily irritations by writing a git plugin.
+  You can do it too!
+published: false
+---
+
+One of the things I do almost every day at my command line is type:
+
+```
+$ git git whatever
+```
+
+It's easy to do.
+
+Sometimes I'll start typing a Git command, but need to look up how to complete
+it, and when I return to the terminal I've forgotten I started typing it and --
+
+```
+$ git git rebase
+git: 'git' is not a git command. See 'git --help'.
+
+Did you mean this?
+	init
+```
+
+Or sometimes I will be inside a [gitsh] shell -- where the `git` prefix is
+implied -- but still type `git` out of habit[^1]. Oops!
+
+  [gitsh]: https://github.com/thoughtbot/gitsh
+
+I've been putting up with this for years, grumbling to myself, "No, I did not
+mean init..." and a few weeks ago decided to do something about it.
+
+The solution turned out to be hiding in the error message I had read so many
+times: `'git' is not a git command`. What if it *was* a git command?
+
+## Git Commands
+
+What actually happens when you run `git status`? For a long time, I imagined
+the `git` command looked something like this:
+
+```ruby
+subcommand = ARGUMENTS.first
+arguments = ARGUMENTS.rest
+
+case subcommand
+when "status": git_status(arguments)
+when "commit": git_commit(arguments)
+# ...
+else
+  abort("'#{subcommand}' is not a git command")
+end
+```
+
+There are many variations on this style, but it's my default method for
```

Comment:
This sentence isn't necessary. It starts me off in a direction about how you write Ruby CLI, but then you go back to the _much more interesting_ way `git` handles subcommands.